flink-test-utils-parent/flink-table-filesystem-test-utils/src/main/java/org/apache/flink/table/file/testutils/catalog/TestFileSystemCatalog.java

```diff
@@ -0,0 +1,687 @@
+    private static final String SCHEMA_FILE_EXTENSION = ".json";
+
+    private final String catalogPathStr;
+
+    private Path catalogPath;
+    private FileSystem fs;
+
+    public TestFileSystemCatalog(String pathStr, String name, String defaultDatabase) {
+        super(name, defaultDatabase);
+        this.catalogPathStr = pathStr;
+    }
+
+    @VisibleForTesting
+    public String getCatalogPathStr() {
+        return catalogPathStr;
+    }
+
+    @Override
+    public Optional<Factory> getFactory() {
+        return Optional.of(new TestFileSystemTableFactory());
+    }
+
+    @Override
+    public void open() throws CatalogException {
+        try {
+            catalogPath = new Path(catalogPathStr);
+            fs = catalogPath.getFileSystem();
+            if (!fs.exists(catalogPath)) {
+                throw new CatalogException(
+                        String.format(
+                                "Catalog %s path %s does not exist.", getName(), catalogPath));
+            }
+            if (!fs.getFileStatus(catalogPath).isDir()) {
+                throw new CatalogException(
+                        String.format(
+                                "Failed to open catalog path. The given path %s is not a directory.",
+                                catalogPath));
+            }
+        } catch (IOException e) {
+            throw new CatalogException(
+                    String.format("Checking catalog path %s exists occur exception.", catalogPath),
+                    e);
+        }
+    }
+
+    @Override
+    public void close() throws CatalogException {}
+
+    @Override
+    public List<String> listDatabases() throws CatalogException {
+        try {
+            FileStatus[] fileStatuses = fs.listStatus(catalogPath);
+            return Arrays.stream(fileStatuses)
+                    .filter(FileStatus::isDir)
+                    .map(fileStatus -> fileStatus.getPath().getName())
+                    .collect(Collectors.toList());
+        } catch (IOException e) {
+            throw new CatalogException("Listing database occur exception.", e);
+        }
+    }
+
+    @Override
+    public CatalogDatabase getDatabase(String databaseName)
+            throws DatabaseNotExistException, CatalogException {
+        if (databaseExists(databaseName)) {
+            return new CatalogDatabaseImpl(Collections.emptyMap(), null);
+        } else {
+            throw new DatabaseNotExistException(getName(), databaseName);
+        }
+    }
+
+    @Override
+    public boolean databaseExists(String databaseName) throws CatalogException {
+        Preconditions.checkArgument(!StringUtils.isNullOrWhitespaceOnly(databaseName));
+
+        return listDatabases().contains(databaseName);
+    }
+
+    @Override
+    public void createDatabase(String name, CatalogDatabase database, boolean ignoreIfExists)
+            throws DatabaseAlreadyExistException, CatalogException {
+        if (databaseExists(name)) {
```

Comment:
Do we need to check if the database name contains "/" like db1/db2/db3 ? Otherwise, it may create subdirectories here.

Reply:
I think there is no need, we should trust the users and leave the choice to them. There may be other unsupported patterns for database names, so we can't enumerate and restrict them.